File: code/dynamic_programming/src/longest_common_subsequence/longest_common_subsequence.java
// Part of Cosmos by OpenGenus Foundation
class LongestCommonSubsequence {
    /* Returns length of longestCommonSubsequence for X[0..m-1], Y[0..n-1] */
  public int longestCommonSubsequence(String text1, String text2) {
        int dp[][] = new int[text1.length() + 1][text2.length() + 1];
        for (int i = 0; i < text1.length(); i++) {
            for (int j = 0; j < text2.length(); j++) {
                if (text1.charAt(i) == text2.charAt(j))
                    dp[i + 1][j + 1] = dp[i][j] + 1;
                else
                    dp[i + 1][j + 1] = Math.max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
        return dp[text1.length()][text2.length()];
    }

  public static void main(String[] args) {
        LongestCommonSubsequence lcs = new LongestCommonSubsequence();
        String s1 = "AGGTAB";
        String s2 = "GXTXAYB";

        System.out.println("Length of LCS is" + " " + lcs.longestCommonSubsequence(s1, s2));
    }
}

File: code/dynamic_programming/src/longest_common_subsequence/longest_common_subsequence.go
/* Part of Cosmos by OpenGenus Foundation */
package main

import "fmt"

/*
Expect Output
The length of LCS of ABCDEFGHIJKLMNOPQRABCEDG adn DEFNMJABCDEG is 9

*/

func max(n1, n2 int) int {
	if n1 > n2 {
		return n1
	}
	return n2
}

func LCS(str1, str2 string) int {
	len1 := len(str1)
	len2 := len(str2)

	dp := make([][]int, len1+1)
	for v := range dp {
		dp[v] = make([]int, len2+1)
	}

	for i := 1; i <= len1; i++ {
		for j := 1; j <= len2; j++ {
			if str1[i-1] == str2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i][j-1], dp[i-1][j])
			}
		}
	}

	return dp[len1][len2]
}

func main() {
	str1 := "ABCDEFGHIJKLMNOPQRABCEDG"
	str2 := "DEFNMJABCDEG"

	lcs := LCS(str1, str2)
	fmt.Printf("The length of LCS of %s adn %s is %d\n", str1, str2, lcs)
}
